frontstage/views/sign_in.py

```diff
@@ -29,27 +29,49 @@ def login():
+            data = {
+                'grant_type': 'password',
+                'client_id': app.config['RAS_FRONTSTAGE_CLIENT_ID'],
+                'client_secret': app.config['RAS_FRONTSTAGE_CLIENT_SECRET'],
+                'username': username,
+                'password': password,
+            }
+            headers ={
+                'Accept': 'application/json',
+                'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
+            }
+
+            oauth2_response = requests.post(url=token_url, data=data , headers=headers, auth=(app.config['RAS_FRONTSTAGE_CLIENT_ID'], app.config['RAS_FRONTSTAGE_CLIENT_SECRET']))
+            oauth2_token = json.loads(oauth2_response.text)
+            # Check to see that this user has not attempted to login too many times or that they have not forgot to
+            # click on the activate account URL in their email by checking the error message back from the OAuth2 server
+            if oauth2_response.status_code == 401:
+                oauth2Error = json.loads(oauth2_response.text)
+                if oauth2Error['detail']== 'Unauthorized user credentials':
+                    return render_template('sign-in/sign-in.html', _theme='default', form=form, data={"error": {"type": "failed"}})
+                elif 'User account locked' in oauth2Error['detail']:
+                    logger.warning('User account is locked on the OAuth2 server')
+                    return render_template('sign-in/sign-in.trouble.html', _theme='default', form=form,
+                                           data={"error": {"type": "account locked"}})
+                elif 'User account not verified' in oauth2Error['detail']:
+                    logger.warning('User account is not verified on the OAuth2 server')
+                    return render_template('sign-in/sign-in.account-not-verified.html', _theme='default', form=form,
+                                           data={"error": {"type": "account not verified"}})
+                else:
+                    logger.error('OAuth 2 server generated 401 which is not understood. Details are: {}'.format(oauth2Error['detail']))
+                    return render_template('sign-in/sign-in.html', _theme='default', form=form,
+                                           data={"error": {"type": "failed"}})
+            if oauth2_response.status_code != 201:
+                logger.error('Unknown error from the OAuth2 server which was: {}'.format(oauth2_response.text))
+                raise ExternalServiceError(oauth2_response.text)
             logger.debug('Access Token Granted')
-
-        except MissingTokenError as e:
-            logger.warning('Missing token', exception=str(e))
-            return render_template('sign-in/sign-in.html', _theme='default', form=form, data={"error": {"type": "failed"}})
+        except (requests.ConnectTimeoutConnectionError, requests.ConnectionError) as e:
```

Comment:
Fair it's not changing anything functional so should be fine
We also have a general error handler which deals with connection errors but that's probably getting changed some time soon